Kava-Labs/webapp
# Fix ledger sequential access errors

- Add a `LedgerMock` class for testing.
  - It uses `setTimeout` to mimic non-blocking IO operations a real ledger has.
  - Non-sequential access will throw an error.
  - It's a bit barebones, but we can improve it over time with an actual test mnemonic to make signing work against a real test chain in CI, dev, etc.
- Adds unit tests for the `CosmosLedger` class that covers the queued operation behavior.
  - `Promise.all` is used to simulate parallel usage of the ledger device across react components
- Manually tested the code against a real device
  - On `master` branch, clicking the tabs fast will throw a sequential access error
  - On this branch, no errors are thrown.
- Add JSDoc
  - Run with `npm run doc`
  - View with `open doc/index.html`
  - Added `CosmosLedger` documentation

TODO before merge:
- [x] add tests to ensure errors go to correct caller (the reject promise path)
- [x] add doc comments to LedgerMock
- [x] update pubkey and signature LedgerMock methods to return correct data type
- [x] add code coverage npm script
